sklearn/naive_bayes.py

```diff
@@ -35,6 +35,30 @@
 __all__ = ['BernoulliNB', 'GaussianNB', 'MultinomialNB']
 
 
+def _safe_logprob(p):
+    """
+    Setting log(0) = -inf as -th where `th` is a very larger number.
+
+    Parameters
+    ----------
+    p : array-like or scalar
+        Probabilities values. Must be within [0, 1] range.
+
+    This avoids summation errors, e.g. np.log(0) - np.log(0) = nan, which
+    may happen while calculating dot product of log probability matrices for
+    joint likelihood estimation
+
+    Note that `th` too large will cause overflow during dot product and wrongly
+    estimate the joint likelihood
+    """
+    th = 1e30
+    p = np.asarray(p)
+    if (p > 1).any() or (p < 0).any():
+        raise ValueError('Input `p` must be within [0, 1] range!')
```

Comment:
if this happens for numerical reasons it does not explain how to make it work. Besides change the data.

why not clipping systematically?